test/soak/serve_hostnames/serve_hostnames.go

```diff
@@ -0,0 +1,202 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+This soak tests places a specified number of pods on each node and then
+repeatedly sends queries to a service running on these pods via
+a serivce
+*/
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+var (
+	queries = flag.Int("queries", 1000, "Number of hostname queries to make in each iteration")
+	perNode = flag.Int("per_node", 1, "Number of hostname pods per node")
+)
+
+const (
+	podStartTimeout = 5 * time.Minute
+)
+
+func main() {
+	flag.Parse()
+
+	settings, err := clientcmd.LoadFromFile(filepath.Join(os.Getenv("HOME"), ".kube", ".kubeconfig"))
+	if err != nil {
+		log.Fatalf("Error loading configuration: %v", err.Error())
+	}
+	config, err := clientcmd.NewDefaultClientConfig(*settings, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		log.Fatalf("Failed to construct config: %v", err)
+	}
+
+	c, err := client.New(config)
+	if err != nil {
+		log.Fatalf("Failed to make client: %v", err)
+	}
+
+	nodes, err := c.Nodes().List()
+	if err != nil {
+		log.Fatalf("Failed to list nodes: %v", err)
+	}
+
+	if len(nodes.Items) == 0 {
+		log.Fatalf("Failed to find any nodes.")
+	}
+
+	fmt.Printf("Nodes found on this cluster:\n")
+	for i, node := range nodes.Items {
+		fmt.Printf("%d: %s\n", i, node.Name)
+	}
+
+	// Make a unique namespace for this test.
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	ns := "serve-hostnames-" + strconv.Itoa(r.Int()%10000)
+	fmt.Printf("Using namespace %s for this test.\n", ns)
+
+	// Put serve-hostname pods on each node.
+	podNames := []string{}
+	for i, node := range nodes.Items {
```

Comment:
Once you hit this loop you should stop using log.Fatalf (because it calls os.Exit and your defer statements won't run).